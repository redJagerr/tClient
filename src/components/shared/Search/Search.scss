.search {
	margin-bottom: toRem(10);
	&__container {
		position: relative;
		display: flex;
		flex-direction: column;
		justify-content: center;
		align-items: center;
	}

	&__city {
		display: flex;
		align-items: center;
		position: relative;
		min-width: 100%;
		height: toRem(280);
		border-radius: 18px;
		overflow: hidden;
		background-color: var(--color-gray);
	}
	&__title {
		position: absolute;
		display: inline-flex;
		flex-direction: column;
		align-items: center;
		justify-content: center;
		min-width: 100%;
		min-height: 64%;
		top: 0;
		left: 0;
		text-align: center;
		vertical-align: middle;
		color: white;
		font-size: toRem(64);
	}
	&__form {
		position: absolute;
		bottom: 14%;
	}
	&__input {
		position: relative;
		min-width: toRem(560);
		min-height: toRem(76);
		padding: toRem(14) toRem(90) toRem(14) toRem(32);
		border-radius: 8px;
		font-size: toRem(16);
		&:focus {
			outline: none;
			border-bottom-left-radius: 0;
			border-bottom-right-radius: 0;
		}
	}
	&__button {
		position: absolute;
		right: toRem(14);
		bottom: toRem(14);
		background-color: var(--primary-color);
		border: none;
		width: toRem(50);
		height: toRem(50);
	}
	&__icon {
		width: 36%;
		height: 36%;
	}
	&__result-list {
		position: absolute;
		display: block;
		background-color: #fff;
		box-shadow: 0.5px 0.5px 5px rgb(183, 183, 183);
		width: 100%;
		border-radius: 8px;
		border-top-left-radius: 0;
		border-top-right-radius: 0;
		z-index: 999;
	}
	&__link {
		position: absolute;
		display: block;
		top: 0;
		left: 0;
		width: 100%;
		height: 100%;
	}
	&__result-item {
		position: relative;
		width: 100%;
		padding: toRem(14) toRem(32);
		border-radius: 8px;
		user-select: none;
		color: black;
		&:hover {
			background-color: var(--color-light-blue);
		}
		&:first-child {
			border-top: 2px rgb(216, 216, 216) solid;
			border-top-left-radius: 0;
			border-top-right-radius: 0;
		}
	}
	@include tablet {
		&__city {
			height: toRem(160);
		}
		&__title {
			min-height: 56%;
			font-size: 7.6vw;
		}
		&__input {
			min-width: 86vw;
			min-height: toRem(56);
			padding-right: toRem(70);
		}
		&__button {
			width: toRem(40);
			height: toRem(40);
			bottom: toRem(8);
			right: toRem(5);
		}
	}
	@include mobile {
		&__city {
			height: toRem(140);
		}
	}
}
