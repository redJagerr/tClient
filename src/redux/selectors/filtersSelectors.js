export const selectFilters = state => state.filters.filters;
export const selectFiltersDistricts = state => state.filters.filters.district;
export const selectFiltersSort = state => state.filters.filters.sort;
export const selectFilterResultTotal = state => state.filterResult.totalPlaces;
export const selectFilterResult = state => state.filterResult;
export const selectFilterSubcategory = state =>
	state.filters.filters.subcategory;
export const selectDistricts = state => state.filters.districts;
export const selectDistrictsLoadStatus = state =>
	state.filters.districtsLoadStatus;
export const selectMobileVisible = state => state.filters.mobileVisible;
export const selectCurrentPage = state => state.filters.filters.currentPage;
export const selectPages = state => state.filterResult.pages;
export const selectSubcategories = state => state.subcategories;
