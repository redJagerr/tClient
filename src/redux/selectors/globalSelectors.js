export const selectDarkTheme = state => state.theme.darkTheme;
export const selectErrorPopupMessage = state => state.global.errorPopupMessage;
export const selectCities = state => state.cities;
