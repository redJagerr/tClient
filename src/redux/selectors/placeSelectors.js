export const selectPlaceData = state => state.place.placeData[0];
export const selectPlaceDataInfo = state => state.place.placeData[0].info;
export const selectPlaceLoadstatus = state => state.place.loadStatus;
export const selectPlaceCategoties = state => state.placeCategories;
