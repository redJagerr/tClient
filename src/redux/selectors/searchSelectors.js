export const selectSearchFilter = state => state.search.searchFilter;
export const selectCityName = state => state.search.cityInfo.name;
export const selectSearch = state => state.search;
