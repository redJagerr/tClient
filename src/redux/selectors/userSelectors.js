export const selectUserInfo = state => state.user.userInfo;
export const selectUserEmail = state => state.user.userInfo.email;
export const selectUserCities = state => state.user.userInfo.cities;
export const selectUserServerError = state => state.user.serverError;
