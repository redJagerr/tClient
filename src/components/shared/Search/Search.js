import { useState, useEffect, useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link, useParams } from 'react-router-dom';
import {
	fetchCityInfo,
	searchRequest,
	updateFilter,
} from 'redux/slices/searchSlice';
import { fetchCities, setIsSearch } from 'redux/slices/citiesSlice';
import { selectDarkTheme } from 'redux/selectors/globalSelectors';
import { selectSearch } from 'redux/selectors/searchSelectors';
import { useOnClickOutside } from 'hooks/use-onclick-outside';
import useDebounce from 'hooks/useDebounce';

import { v4 } from 'uuid';
import { Skeleton } from '@mui/material';
import searchIcon from 'assets/pictures/svg/search.svg';

const Search = ({ isMain, searchPlaceholder }) => {
	const [value, setValue] = useState('');
	const [opened, setOpened] = useState(false);
	const debouncedValue = useDebounce(value, 300);
	const { cityId } = useParams();
	const dispatch = useDispatch();
	const inputRef = useRef(null);
	const resultsRef = useRef(null);

	const { cityInfo, searchResult, searchFilter, loadStatus } =
		useSelector(selectSearch);

	const darkTheme = useSelector(selectDarkTheme);

	useEffect(() => {
		if (!isMain) dispatch(fetchCityInfo(cityId));
		if (searchFilter !== '') dispatch(updateFilter(''));
	}, []);

	useEffect(() => {
		if (debouncedValue === '') return;
		isMain
			? dispatch(searchRequest(`?search=${debouncedValue}`))
			: dispatch(searchRequest(`${cityId}/place/${debouncedValue}`));
	}, [debouncedValue]);

	useOnClickOutside(resultsRef, () => setOpened(false), !opened, inputRef);

	const renderSearchResult = () =>
		debouncedValue !== ''
			? searchResult.map(i => (
					<li key={v4()} className='search__result-item'>
						<Link
							to={isMain ? `/${i._id}/places` : `${i.id}/info`}
							className='link search__link'
							onClick={() => setValue('')}
						></Link>
						{i.name}
					</li>
			  ))
			: null;

	const addSearchFilter = e => {
		e.preventDefault();
		if (isMain) {
			dispatch(fetchCities(`?search=${value}`));
			dispatch(setIsSearch(value === '' ? false : true));
			return;
		}
		dispatch(updateFilter(value));
	};

	const cityImage = isMain
		? 'https://s1.1zoom.ru/big3/804/Russia_Moscow_Rivers_478099.jpg'
		: cityInfo.photos;
	const cityName = isMain ? '' : cityInfo.name;

	return (
		<section className='search'>
			<div className='search__container'>
				<div className={`search__city ${darkTheme ? 'dark-card' : ''}`}>
					<h2 className='search__title'>
						{loadStatus === 'loading' ? <Skeleton width={300} /> : cityName}
					</h2>
					{loadStatus === 'loading' ? null : (
						<img src={cityImage} alt='City view' className='image' />
					)}
				</div>
				<form action='#' className='search__form'>
					<input
						ref={inputRef}
						type='text'
						className={`input search__input`}
						placeholder={searchPlaceholder}
						value={value}
						onInput={e => setValue(e.target.value)}
						onFocus={() => setOpened(true)}
					/>
					<button
						className='button search__button'
						onClick={e => addSearchFilter(e)}
					>
						<img className='search__icon' alt='search icon' src={searchIcon} />
					</button>

					{opened ? (
						<ul ref={resultsRef} className='list search__result-list'>
							{renderSearchResult()}
						</ul>
					) : null}
				</form>
			</div>
		</section>
	);
};
export default Search;
