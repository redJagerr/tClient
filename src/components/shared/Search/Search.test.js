import Search from './Search';
import * as reduxHooks from 'react-redux';
import { renderWithRoutes } from 'tests/helpers/renderWithRoutes';
import { selectDarkTheme } from 'redux/selectors/globalSelectors';
import { selectSearch } from 'redux/selectors/searchSelectors';
import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

jest.mock('react-redux');
jest.mock('redux/selectors/searchSelectors');
jest.mock('redux/selectors/globalSelectors');

const mockState = {
	search: {
		searchFilter: '',
		cityInfo: {
			name: 'name',
		},
		searchResult: [],
		opened: false,
		loadStatus: 'idle',
		searchLoadStatus: 'idle',
	},
	theme: {
		darkTheme: false,
	},
};
describe('FilterResult component', () => {
	beforeEach(() => {
		const mockUseSelector = jest
			.spyOn(reduxHooks, 'useSelector')
			.mockImplementation(cb => cb(mockState));
		let mockUseDispatch = jest.spyOn(reduxHooks, 'useDispatch');
		const dispatch = jest.fn();
		mockUseDispatch.mockReturnValue(dispatch);
		selectDarkTheme.mockReturnValue(mockState.theme.darkTheme);
		selectSearch.mockReturnValue(mockState.search);
	});
	it('should match with snapshot', () => {
		const component = renderWithRoutes(<Search />);
		expect(component).toMatchSnapshot();
	});

	it('should change value after user input', () => {
		renderWithRoutes(<Search />);
		const input = screen.getByRole('textbox');
		userEvent.type(input, 'text');
		expect(input.value).toBe('text');
	});
});
