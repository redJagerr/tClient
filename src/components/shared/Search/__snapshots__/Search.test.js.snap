// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterResult component should match with snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section
        class="search"
      >
        <div
          class="search__container"
        >
          <div
            class="search__city "
          >
            <h2
              class="search__title"
            >
              name
            </h2>
            <img
              alt="City view"
              class="image"
            />
          </div>
          <form
            action="#"
            class="search__form"
          >
            <input
              class="input search__input"
              type="text"
              value=""
            />
            <button
              class="button search__button"
            >
              <img
                alt="search icon"
                class="search__icon"
                src="search.svg"
              />
            </button>
          </form>
        </div>
      </section>
    </div>
  </body>,
  "container": <div>
    <section
      class="search"
    >
      <div
        class="search__container"
      >
        <div
          class="search__city "
        >
          <h2
            class="search__title"
          >
            name
          </h2>
          <img
            alt="City view"
            class="image"
          />
        </div>
        <form
          action="#"
          class="search__form"
        >
          <input
            class="input search__input"
            type="text"
            value=""
          />
          <button
            class="button search__button"
          >
            <img
              alt="search icon"
              class="search__icon"
              src="search.svg"
            />
          </button>
        </form>
      </div>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
